### PR TITLE
Translate Objectif into CEG

### PR DESCRIPTION
**Pull request by fischJan**
I already opened a pull request in our old repo and got feedback by max (see pull request #471):

1) match line breaks
2) add a sensible exception message that xtext parsing failed
3) enable umlauts
4) add "Sonst" effects to your grammar and visualize it within a CEG

all comments have been incorporated into the new commits

### CODE REVIEW

File: bundles/specmate-integration-test/src/com/specmate/test/integration/ObjectifTest.java [hunk 1]
```diff
@@ -18,95 +23,267 @@
...
+
+		ObjectifNode effectB = ruleB.getEffect();
+		Assert.assertTrue(effectB instanceof LiteralNode);
+		Assert.assertTrue(((LiteralNode) effectB).getContent().equals("C=1"));
+
+		Assert.assertTrue(ruleB.hasAlternative());
+		ObjectifNode alternative = ruleB.getAlternative();
+		Assert.assertTrue(alternative instanceof LiteralNode);
+		Assert.assertTrue(((LiteralNode) alternative).getContent().equals("C=2"));
+
```
Comment (tobi321) [inline bundles/specmate-integration-test/src/com/specmate/test/integration/ObjectifTest.java]: Remove empty line

File: bundles/specmate-integration-test/src/com/specmate/test/integration/ObjectifTest.java [hunk 2]
```diff
@@ -18,95 +23,267 @@
 import com.specmate.objectif.resolve.rule.OrNode;
 import com.specmate.xtext.XTextException;
 
-public class ObjectifTest {
+// Two types of tests:
```
Comment (tobi321) [inline bundles/specmate-integration-test/src/com/specmate/test/integration/ObjectifTest.java]: Change the comment type to  /***/:
`/** Two types of tests:  1) tests for checking the functionality of the xtext parsing 2) tests for checking the functionality of the pseudo code translator */`

File: bundles/specmate-model-generation/generated/buildfiles [hunk 3]
```diff
@@ -0,0 +1 @@
+A:/specmate/bundles/specmate-model-generation/generated/specmate-model-generation.jar
```
Comment (tobi321) [inline bundles/specmate-model-generation/generated/buildfiles:1]: Do we need this buildfile?

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+import com.specmate.model.requirements.NodeType;
+import com.specmate.objectif.resolve.BusinessRuleUtil;
+import com.specmate.objectif.resolve.rule.AndNode;
+import com.specmate.objectif.resolve.rule.BusinessRuleNode;
+import com.specmate.objectif.resolve.rule.LiteralNode;
+import com.specmate.objectif.resolve.rule.ObjectifNode;
+import com.specmate.objectif.resolve.rule.OrNode;
+import com.specmate.xtext.XTextException;
+
+// This class transforms pseudo code into a CEG.
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+		} catch (XTextException e) {
+			throw new SpecmateInternalException(ErrorCode.INTERNAL_PROBLEM,
+					"Unfortunately, the XText parsing of your inserted pseudo code went wrong." + e.getMessage());
+		}
+
+		BusinessRuleNode rule = rules.get(0);
+		return createCEGModelForEachSection(rule, model, null, false);
+	}
+
+	// This method creates a CEG model for each "WENN" --> "DANN" --> "ENDE-WENN"
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** */ comment

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+	// clause.
+	private CEGModel createCEGModelForEachSection(BusinessRuleNode rule, CEGModel model, CEGNode brNode,
+			boolean nesting) {
+
+		// The type of a BR node should be always "OR". Exception: The model consists
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+		if (brNode == null) {
+			cegNodeBR = creation.createNode(model, "BR" + (brnum++), "is present", 0, 0, NodeType.OR);
+		} else {
+			// Otherwise, we are within a recursive loop (i.e. the clause is nested) and the
+			// brNode is our superior
+			// business rule.
+			cegNodeBR = brNode;
+		}
+
+		// Check for nesting:
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+		// The rootEffect contains all effects of a rule. If the rootEffect is a
+		// "BusinessRuleNode", the rule is nested and consists of multiple rules.
+		// Here we check if we need to do a recursion again or if we reached the
+		// business rule at the lowest depth (i.e. the business rule which does not
+		// contains another one).
+		ObjectifNode rootEffect = rule.getEffect();
+		if (rootEffect instanceof BusinessRuleNode) {
+			// Case: nested
+			// Basic idea: Split the whole tree in subtrees (i.e. split the tree into
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+			// superior rule
+			if (nesting == false) {
+				connectCausesToBR(cegNodeBR, singleOrCauses, andCausesGroups, model);
+			} else {
+				// inferior rule
+				connectCausesToNestedBR(cegNodeBR, singleOrCauses, andCausesGroups, model);
+			}
+
+			// We do not need to identify the effects of the business rule since we know
+			// that the rule contains another one and is therefore a condition for the
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+			// nested rule. We simply connect the rule with the inferior rule.
+			creation.createConnection(model, cegNodeBR, cegNodeBRNested, false);
+
+			// Start with the recursion. We update the root and do the whole procedure for
+			// the inferior rule again. Until we reach a rule which is not nested anymore.
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+			createCEGModelForEachSection((BusinessRuleNode) rootEffect, model, cegNodeBRNested, true);
+
+		} else {
+			// Case: not nested
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+			List<ObjectifNode> alternativeEffects = new ArrayList<ObjectifNode>();
+			
+			identifyEffects(rootEffect, effects);
+
+			if (rule.hasAlternative()) {
+				rootEffect = rule.getAlternative();
+				identifyEffects(rootEffect, alternativeEffects);
+			}
+
+			// 1. Case = The rule contains only one cause. This is the case if the rootCause
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+			// is a "LiteralNode".
+			if (rootCause instanceof LiteralNode) {
+				// Create a node with the content of the rootCause and connect the node with the
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+				// rule node by means of identity connection.
+				CEGNode cegNode = creation.createNode(model, ((LiteralNode) rootCause).getContent(), "is present", 0, 0,
+						null);
+
+				creation.createConnection(model, cegNode, cegNodeBR, false);
+				// Search for effects of the rule.
+				//identifyEffects(rootEffect, effects);
+
+//				if (rule.hasAlternative()) {
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: commented out code is duplicate of line 110-113. Remove.

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 15]
```diff
@@ -0,0 +1,378 @@
...
+
+				// Connect the rule node with its alternative effects
+				for (ObjectifNode effect : alternativeEffects) {
+					CEGNode cegNodeEffect = creation.createNode(model, ((LiteralNode) effect).getContent(),
+							"is present", 0, 0, null);
+					creation.createConnection(model, cegNodeBR, cegNodeEffect, true);
+				}
+			} else {
+
+				// 2. Case = The rule contains multiple causes which are interconnected
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+				}
+				connectBRToEffects(cegNodeBR, effects, alternativeEffects, model);
+			}
+		}
+		return model;
+	}
+
+	private void connectBRToEffects(CEGNode cegNodeBR, List<ObjectifNode> effects,
+			List<ObjectifNode> alternativeEffects, CEGModel model) {
+		// Iterate over all effects and connect the business rule node with these
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+
+				creation.createConnection(model, intermediateNode, cegNodeBR, false);
+			}
+
+			for (ObjectifNode cause : andCauses) {
+				CEGNode cegNode = creation.createNode(model, ((LiteralNode) cause).getContent(), "is present", 0, 0,
+						null);
+
+				// There are two cases:
+				// a) there is only one set of conjunctive causes and no single / disjunctive
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+				// causes --> just connect the causes with the rule without intermediate node.
+				if (andCausesGroups.size() == 1 && singleOrCauses.size() == 0) {
+					creation.createConnection(model, cegNode, cegNodeBR, false);
+					cegNodeBR.setType(NodeType.AND);
+				} else {
+					// b) otherwise we need an intermediate node and connect conjunctive causes to
+					// these nodes. All related causes share the same intermediate node!
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 19]
```diff
@@ -0,0 +1,378 @@
...
+					creation.createConnection(model, cegNode, intermediateNode, false);
+				}
+			}
+		}
+		// We iterate over all single causes and connect them with the BR by means of a
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 20]
```diff
@@ -0,0 +1,378 @@
...
+		// disjunction.
+		for (ObjectifNode cause : singleOrCauses) {
+			CEGNode cegNode = creation.createNode(model, ((LiteralNode) cause).getContent(), "is present", 0, 0, null);
+			creation.createConnection(model, cegNode, cegNodeBR, false);
+		}
+	}
+
+	// Implementation for nested BRs:
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 21]
```diff
@@ -0,0 +1,378 @@
...
+				connectAndCauses(rootEffect.getLeft(), effects);
+			}
+
+			if (rootEffect.getRight() instanceof AndNode) {
+				connectAndCauses(rootEffect.getRight(), effects);
+			}
+		}
+	}
+
+	// Here we connect causes which are interconnected using conjunctions.
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 22]
```diff
@@ -0,0 +1,378 @@
...
+
+			if (rootCause.getRight() instanceof AndNode) {
+				// recursive call of the method but with updated rootcause.
+				// See explanation above.
+				connectAndCauses(rootCause.getRight(), andCauses);
+			}
+		}
+	}
+
+	// Here we separate between isolated causes (disjunctive causes) and related
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java [hunk 23]
```diff
@@ -0,0 +1,378 @@
...
+				orCauses.add(rootCause.getLeft());
+			}
+
+			// belongs to case 1
+			if (rootCause.getRight() instanceof LiteralNode) {
+				orCauses.add(rootCause.getRight());
+			}
+
+			if (rootCause.getLeft() instanceof AndNode) {
+				// Here we identify a conjunction. We create a list of Objectif nodes for each
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromPseudoCode.java]: Change to /** Comment */

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromRequirementService.java [hunk 24]
```diff
@@ -89,7 +89,13 @@ private CEGModel generateModelFromDescription(CEGModel model) throws SpecmateExc
...
 		ICEGFromRequirementGenerator generator;
 		
-		generator = new PatternbasedCEGGenerator(lang, tagger, this.configService); 
+
+		if(lang == ELanguage.PSEUDO) {
+			generator = new GenerateModelFromPseudoCode();
+		} else {
+			generator = new PatternbasedCEGGenerator(lang, tagger, this.configService); 
+		}
+
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromRequirementService.java:97]: Remove empty line

File: bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromRequirementService.java [hunk 25]
```diff
@@ -89,7 +89,13 @@ private CEGModel generateModelFromDescription(CEGModel model) throws SpecmateExc
 		ELanguage lang = NLPUtil.detectLanguage(text);
 		ICEGFromRequirementGenerator generator;
 		
-		generator = new PatternbasedCEGGenerator(lang, tagger, this.configService); 
+
```
Comment (tobi321) [inline bundles/specmate-model-generation/src/com/specmate/modelgeneration/GenerateModelFromRequirementService.java]: Remove empty line

**Review by tobi321 — CHANGES_REQUESTED** [review]
Good job!
Works in old editor, merge develop into branch and check out if it still works with mx-graph. 
In old editor, all nodes are on the top left hiding each other. Is this the expected behaviour? 
Change the comments to multiline style (see below).

**Comment by fischJan** [comment]
> Good job!
> Works in old editor, merge develop into branch and check out if it still works with mx-graph.
> In old editor, all nodes are on the top left hiding each other. Is this the expected behaviour?
> Change the comments to multiline style (see below).

**Review by tobi321 — APPROVED** [review]
Good job!

**Pull request merged**